src/server/pachw/server/pachw.go

```
@@ -0,0 +1,117 @@
+	ctx = auth.AsInternalUser(ctx, "pachw-controller")
+	backoff.RetryUntilCancel(ctx, func() (retErr error) { //nolint:errcheck
+		lock := dlock.NewDLock(p.env.EtcdClient, path.Join(p.env.EtcdPrefix, "pachw-controller-lock"))
+		defer func() {
+			if err := lock.Unlock(ctx); err != nil {
+				unlockErr := errors.Wrap(err, "error unlocking in pachw controller")
+				p.log.Error(unlockErr.Error())
+				retErr = multierror.Append(retErr, unlockErr)
+			}
+		}()
+		var replicas int
+		var scaleDownCount int
+		ticker := time.NewTicker(period)
+		defer ticker.Stop()
+		for {
+			numTasks := p.countTasks(ctx, []string{server.URLTaskNamespace, server.StorageTaskNamespace})
+			scaleDownCount++
+			if numTasks > replicas/2 {
+				scaleDownCount = 0
+			}
+			if numTasks > replicas {
+				replicas = miscutil.Min(numTasks, p.env.MaxReplicas)
+				p.log.Infof("scaling up pachw workers to %d replicas", replicas)
+				if err := p.scalePachw(ctx, int32(replicas)); err != nil {
+					p.log.Errorf("failed to scale up pachw: %v", err)
+					return errors.EnsureStack(err)
+				}
+			} else if scaleDownCount == scaleDownThreshold {
+				replicas = miscutil.Max(replicas/2, p.env.MinReplicas)
+				scaleDownCount = 0
+				p.log.Infof("scaling down pachw workers to %d replicas", replicas)
+				if err := p.scalePachw(ctx, int32(replicas)); err != nil {
+					p.log.Errorf("failed to scale down pachw: %v", err)
+					return errors.EnsureStack(err)
+				}
+			}
+			select {
+			case <-ticker.C:
+			case <-ctx.Done():
+				return errors.EnsureStack(ctx.Err())
+			}
+		}
+	}, backoff.NewInfiniteBackOff(), func(err error, _ time.Duration) error {
+		p.log.Errorf("error in pachw run: %v", err)
+		return nil
+	})
+}
+
+func (p *pachW) countTasks(ctx context.Context, namespaces []string) int {
+	totalTasks := 0
+	for _, ns := range namespaces {
+		tasks, _, err := task.Count(ctx, p.env.TaskService, ns, "")
+		if err != nil {
+			p.log.Errorf("error counting tasks: %v", err)
```

We should just return an error here.